File: _CONTAINER/_BY_MONTH/_Feb/Thesis-physics-master/doc/particles/particleEmitter.cpp
/**
 * @fileoverview Particle emitter.
 * @author sebastian.poreba@gmail.com (Sebastian Poreba)
 */
#include "particleEmitter.h"

smash::ParticleEmitter::ParticleEmitter() {
  positionX = 0;
  positionY = 0;
  angle = 0;
  velocity = 10;
  velocitySpread = 0.2;
  spread = M_PI * 10 / 180;
  lifespan = 50;
  productionRate = 10;
};


/**
 * @param angle
 */
void smash::ParticleEmitter::setAngle(float angle) {
  this->angle = M_PI * angle / 180;
};


/**
 * @param velocity
 */
void smash::ParticleEmitter::setVelocity(float velocity) {
  this->velocity = velocity;
};


/**
 * @param velocitySpread
 */
void smash::ParticleEmitter::setVelocitySpread(float velocitySpread) {
  this->velocitySpread = velocitySpread;
};


/**
 * @param spread
 */
void smash::ParticleEmitter::setSpread(float spread) {
  this->spread = M_PI * spread / 180;
};


/**
 * @param lifespan
 */
void smash::ParticleEmitter::setLifespan(float lifespan) {
  this->lifespan = lifespan;
};


/**
 * @param rate
 */
void smash::ParticleEmitter::setProductionRate(int rate) {
  this->productionRate = rate;
};


/**
 * @return {!Array.<!smash::Particle>}
 */
std::vector<smash::Particle*>* smash::ParticleEmitter::getNewParticles() {
  std::vector<smash::Particle*> *newParticles =
      new std::vector<smash::Particle*>;
  for (int i = 0; i < this->productionRate; i++) {
    smash::Particle* p = new smash::Particle();
    p->lifespan = this->lifespan;
    p->positionX = this->positionX;
    p->positionY = this->positionY;
    p->velocityX = sin(this->angle +
        (((float) rand() / (RAND_MAX)) - 0.5) * this->spread) *
        this->velocity *
        (1 + (((float) rand() / (RAND_MAX)) - 0.5) * this->velocitySpread);
    p->velocityY = cos(this->angle +
        (((float) rand() / (RAND_MAX)) - 0.5) * this->spread) *
        this->velocity *
        (1 + (((float) rand() / (RAND_MAX)) - 0.5) * this->velocitySpread);
    newParticles->push_back(p);
  }
  return newParticles;
};

File: _CONTAINER/_BY_MONTH/_Feb/Thesis-physics-master/browser/engine/particleEmitter.js
/**
 * @fileoverview Particle emitter.
 * @author sebastian.poreba@gmail.com (Sebastian Poreba)
 */

goog.provide('smash.ParticleEmitter');

goog.require('smash.Particle');



/**
 * @constructor
 */
smash.ParticleEmitter = function() {
  /**
   * @type {number}
   */
  this.positionX = 0.1;

  /**
   * @type {number}
   */
  this.positionY = 0.1;

  /**
   * @type {number}
   */
  this.angle = 0.1;

  /**
   * @type {number}
   */
  this.velocity = 10.1;

  /**
   * @type {number}
   */
  this.velocitySpread = 0.2;

  /**
   * @type {number}
   */
  this.spread = Math.PI * 10 / 180;

  /**
   * In ticks.
   * @type {number}
   */
  this.lifespan = 50;

  /**
   * @type {number}
   */
  this.productionRate = 10;
};


/**
 * @param {number} angle
 */
smash.ParticleEmitter.prototype.setAngle = function(angle) {
  this.angle = Math.PI * angle / 180;
};


/**
 * @param {number} velocity
 */
smash.ParticleEmitter.prototype.setVelocity = function(velocity) {
  this.velocity = velocity;
};


/**
 * @param {number} velocitySpread
 */
smash.ParticleEmitter.prototype.setVelocitySpread = function(velocitySpread) {
  this.velocitySpread = velocitySpread;
};


/**
 * @param {number} spread
 */
smash.ParticleEmitter.prototype.setSpread = function(spread) {
  this.spread = Math.PI * spread / 180;
};


/**
 * @param {number} lifespan
 */
smash.ParticleEmitter.prototype.setLifespan = function(lifespan) {
  this.lifespan = lifespan;
};


/**
 * @param {number} rate
 */
smash.ParticleEmitter.prototype.setProductionRate = function(rate) {
  this.productionRate = rate;
};


/**
 * @return {!Array.<!smash.Particle>}
 */
smash.ParticleEmitter.prototype.getNewParticles = function() {
  var newParticles = [];
  for (var i = 0; i < this.productionRate; i++) {
    var p = new smash.Particle();
    p.lifespan = this.lifespan;
    p.positionX = this.positionX;
    p.positionY = this.positionY;
    p.velocityX = Math.sin(this.angle +
        (Math.random() - 0.5) * this.spread) *
        this.velocity * this.velocitySpread;
    p.velocityY = Math.cos(this.angle +
        (Math.random() - 0.5) * this.spread) *
        this.velocity *
        (1 + (Math.random() - 0.5) * this.velocitySpread);
    newParticles.push(p);
  }
  return newParticles;
};
